### DISCUSSION TITLE
CHDMISoundBaseDevice feature extension up to 8ch.

### DISCUSSION THREAD

**Discussion by Newdrake** [rsta2/circle#477] [Development]
May be useful for professional audio application the extension of the audio channels of the circle class CHDMISoundBaseDevice from actual limits of 2ch up to 8ch.
Trying to implement this feature extension I've searched for the necessary changes to do.
There are many define in the code fixing this limit, SOUND_HW_CHANNELS WRITE_CHANNELS and also some value in the SetAudioInfoFrame method, but also modifying these value to the new expect correct value the driver doesn't work correctly.
On the other side, it seem not available the technical documentation of the HDMI internal peripheral of BCM2711 to have a clear way on how to set the correct parameters in the registers to upgrade the number of channels.
There is also a Linux implementation of this driver working up to 8ch but the implementation is completely different and not very easy to derive from this the modification needed in the Circle code.
I am asking for suggestion for the right strategy to do this.

**Comment by rsta2** [comment]
You could do an insensitive search for "chan" in the file drivers/gpu/drm/vc4/
vc4_hdmi.c in Linux and look for the same registers in the Circle driver, when 
you have a match. But there is also the HDMI display driver in this file, so 
not all matches will be important for HDMI audio. I guess you have to modify 
soundbasedevice.cpp too, where SoundFormatIEC958 is found. You must use the 
standard constructor of CSoundBaseDevice in CHDMISoundBaseDevice and 
explicitly call CSoundBaseDevice::Setup() to be able to use 8 channels.

**Reply by Newdrake** [reply]
Thanks a lot for your useful guiding indication, now the CHDMISoundBaseDevice driver is working with 8 channels!
In my application I need two instance of the CSoundBaseDevice class, one for I2S and one for HDMI, but when using the .write method of HDMI instance the audio system collapse, there could be any conflict between this two instances at m_Interrupt or DMA level?

**Reply by rsta2** [reply]
Great, you got it working with 8 channels! I'm not aware of a conflict, but you will have twice the IRQs in a time interval.

**Comment by Newdrake** [comment]
The CHDMISoundBaseDevice driver can be configured to be used as input HDMI?
On the CM4 there are two HDMI port, is possible to use two instance of the class, one configured as audio out and the other as audio input?

**Reply by rsta2** [reply]
No, the class `CHDMISoundBaseDevice` works only in output (write) direction.

**Comment by Newdrake** [comment]
In the CHDMISoundBaseDevice class driver there are many methods for the input:
CSoundBaseDevice::AllocateReadQueue
CSoundBaseDevice::SetReadFormat
CSoundBaseDevice::Read
CSoundBaseDevice::GetReadQueueSizeFrames
..

Are these method working on the HDMI interface of CM4 for read audio data from HDMI port?
It seem the HDMI0 and HDMI1 ports of CM4 configured as output instead.
Thanks a lot for your support.

**Reply by rsta2** [reply]
On the RPi 4 only HDMI0 is supported. This be should the case on CM4 too.

**Comment by Newdrake** [comment]
ok!
The CUSBSoundBaseDevice class driver can be used in DeviceModeRXOnly, m_nBitResolution=24 and m_nSampleRate=44100Hz?

**Comment by rsta2** [comment]
Yes, you have to set the following option in the first line of the file cmdline.txt on the SD card to 
use 24-bit samples:

soundopt=24

**Comment by Newdrake** [comment]
thanks a lot!

**Reply by rsta2** [reply]
You are welcome. BTW [these devices](https://github.com/rsta2/circle/wiki/Device-compatibility-list#usb-audio-streaming-devices) have been successfully tested with the Circle USB audio streaming driver. More should work, but these devices can be very different, so it's not safe.

**Comment by Newdrake** [comment]
I'm trying to use the chip CM108 from C-Media electronics, this is a converter from I2S to USB audio format, have you some feedback about compatibility of this with Circle driver?

**Reply by rsta2** [reply]
Unfortunately there hasn't been feedback on this.
